server/service/schedule/schedule.go

```diff
@@ -0,0 +1,292 @@
+	return func(s *Schedule) {
+		s.configReloadInterval = interval
+		s.configReloadIntervalFn = fn
+	}
+}
+
+// WithAltLockID sets an alternative identifier to use when acquiring the lock.
+//
+// If not set, then the Schedule's name is used for acquiring the lock.
+func WithAltLockID(name string) Option {
+	return func(s *Schedule) {
+		s.altLockName = name
+	}
+}
+
+// WithJob adds a job to the Schedule.
+//
+// Each job is executed in the order they are added.
+func WithJob(id string, fn JobFn) Option {
+	return func(s *Schedule) {
+		s.jobs = append(s.jobs, Job{
+			ID: id,
+			Fn: fn,
+		})
+	}
+}
+
+// New creates and returns a Schedule.
+// Jobs are added with the WithJob Option.
+//
+// The jobs are executed serially in order at the provided interval.
+//
+// The provided locker is used to acquire/release a lock before running the jobs.
+// The provided name and instanceID of the Schedule is used as the locking identifier.
+func New(
+	ctx context.Context,
+	name string,
+	instanceID string,
+	interval time.Duration,
+	locker Locker,
+	opts ...Option,
+) *Schedule {
+	sch := &Schedule{
+		ctx:                  ctx,
+		name:                 name,
+		instanceID:           instanceID,
+		logger:               log.NewNopLogger(),
+		done:                 make(chan struct{}),
+		configReloadInterval: 1 * time.Hour, // by default we will check for updated config once per hour
+		schedInterval:        interval,
+		locker:               locker,
+	}
+	for _, fn := range opts {
+		fn(sch)
+	}
+	return sch
+}
+
+// Start starts running the added jobs.
+//
+// All jobs must be added before calling Start.
+func (s *Schedule) Start() {
+	var m sync.Mutex // protects currentStart and currentWait.
+	currentStart := time.Now()
+	currentWait := 10 * time.Second
+
+	getWaitTimes := func() (start time.Time, wait time.Duration) {
+		m.Lock()
+		defer m.Unlock()
+
+		return currentStart, currentWait
+	}
+
+	setWaitTimes := func(start time.Time, wait time.Duration) {
+		m.Lock()
+		defer m.Unlock()
+
+		currentStart = start
+		currentWait = wait
+	}
+
+	if schedInterval := s.getSchedInterval(); schedInterval < currentWait {
+		setWaitTimes(currentStart, schedInterval)
+	}
+
+	var g sync.WaitGroup
+
+	schedTicker := time.NewTicker(currentWait)
+	g.Add(+1)
+	go func() {
+		defer g.Done()
+
+		for {
+			_, currWait := getWaitTimes()
+			level.Debug(s.logger).Log("msg", "waiting", "current wait time", currWait)
+
+			select {
+			case <-s.ctx.Done():
+				return
+
+			case <-schedTicker.C:
+				level.Debug(s.logger).Log("waiting", "done")
+
+				schedInterval := s.getSchedInterval()
+				schedTicker.Reset(schedInterval)
+
+				newStart := time.Now()
+				newWait := schedInterval
+				setWaitTimes(newStart, newWait)
+
+				if ok := s.acquireLock(); !ok {
+					continue
+				}
+
+				for _, job := range s.jobs {
+					level.Debug(s.logger).Log("msg", "starting", "jobID", job.ID)
+					if err := job.Fn(s.ctx); err != nil {
```

Comment:
There's no panic recovery, so a buggy job could bring down the scheduler. For a generic task runner like this, it might be worth implementing that recovery?